domain.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package gomaasapi
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/schema"
+	"github.com/juju/version"
+)
+
+type domain struct {
+	authoritative       bool
+	resourceRecordCount int
+	ttl                 int
+	resourceURI         string
+	id                  int
+	name                string
+}
+
+// implements Domain
+func (domain *domain) Name() string {
+	return domain.name
+}
+
+func readDomains(controllerVersion version.Number, source interface{}) ([]*domain, error) {
+	checker := schema.List(schema.StringMap(schema.Any()))
+	coerced, err := checker.Coerce(source, nil)
+	if err != nil {
+		return nil, errors.Annotatef(err, "domain base schema check failed")
+	}
+	valid := coerced.([]interface{})
+	return readDomainList(valid)
+}
+
+func domain_(source map[string]interface{}) (*domain, error) {
+	fields := schema.Fields{
+		"authoritative":         schema.Bool(),
+		"resource_record_count": schema.ForceInt(),
+		"ttl":          schema.OneOf(schema.Nil("null"), schema.ForceInt()),
+		"resource_uri": schema.String(),
+		"id":           schema.ForceInt(),
+		"name":         schema.String(),
+	}
+	checker := schema.FieldMap(fields, nil) // no defaults
+	coerced, err := checker.Coerce(source, nil)
+	if err != nil {
+		return nil, errors.Annotatef(err, "domain schema check failed")
+	}
+	valid := coerced.(map[string]interface{})
+
+	// ttl live is going to be the zero value if it is null. Thus the time
+	// to live will be zero. Would rather user's check for zero or for nil?
+	ttl, _ := valid["ttl"].(int)
```

Comment:
I do wonder if we should just mirror the MAAS behaviour and have *int.

Reply:
Just to clarify, TTL = 0 has a valid meaning in DNS (it means don't cache) rather than no TTL meaning use the default value for caching.

I don't think Juju is using the TTL right now, but we should track the data correctly.